File: Python/CreationalDesignPattern/FactoryMethodDesignpattern.py
from abc import ABC, abstractmethod


# Product Interface
class INotification(ABC):
    @abstractmethod
    def send(self):
        pass


class SMSNotification(INotification):
    def send(self):
        raise NotImplementedError("Unimplemented method 'send'")


class WhatsAppNotification(INotification):
    def send(self):
        raise NotImplementedError("Unimplemented method 'send'")


class INotificationFactory(ABC):
    @abstractmethod
    def get_notification(self) -> INotification:
        pass


class SMSNotificationFactory(INotificationFactory):
    def get_notification(self) -> INotification:
        return SMSNotification()


class WhatsAppNotificationFactory(INotificationFactory):
    def get_notification(self) -> INotification:
        return WhatsAppNotification()


class Slack(INotification):
    def send(self):
        pass


class SlackNotificationFactory(INotificationFactory):
    def get_notification(self) -> INotification:
        return Slack()


class NotificationService:
    def send_notification(self, factory: INotificationFactory):
        notification = factory.get_notification()
        notification.send()


def main():
    service = NotificationService()
    service.send_notification(SMSNotificationFactory())


if __name__ == "__main__":
    main()

File: C++/CreationalDesignPattern/FactoryMethodDesignpattern.cpp
#include <memory>
#include <stdexcept>

// Product Interface
struct INotification {
    virtual void send() = 0;
    virtual ~INotification() = default;
};

struct SMSNotification : INotification {
    void send() override {
        throw std::runtime_error("Unimplemented method 'send'");
    }
};

struct WhatsAppNotification : INotification {
    void send() override {
        throw std::runtime_error("Unimplemented method 'send'");
    }
};

struct INotificationFactory {
    virtual std::unique_ptr<INotification> getNotification() = 0;
    virtual ~INotificationFactory() = default;
};

struct SMSNotificationFactory : INotificationFactory {
    std::unique_ptr<INotification> getNotification() override {
        return std::make_unique<SMSNotification>();
    }
};

struct WhatsAppNotificationFactory : INotificationFactory {
    std::unique_ptr<INotification> getNotification() override {
        return std::make_unique<WhatsAppNotification>();
    }
};

struct Slack : INotification {
    void send() override {}
};

struct SlackNotificationFactory : INotificationFactory {
    std::unique_ptr<INotification> getNotification() override {
        return std::make_unique<Slack>();
    }
};

class NotificationService {
public:
    void sendNotification(INotificationFactory& factory) {
        auto notification = factory.getNotification();
        notification->send();
    }
};

int main() {
    NotificationService service;
    SMSNotificationFactory factory;
    service.sendNotification(factory);
    return 0;
}
